ext/json/ext/generator/generator.c

```diff
@@ -165,6 +166,84 @@ static const unsigned char script_safe_escape_table[256] = {
      4, 4, 4, 4, 4, 4, 4, 4, 5, 5, 5, 5, 6, 6, 9, 9,
 };
 
+
+typedef struct _search_state {
+    const char *ptr;
+    const char *end;
+    const char *cursor;
+    FBuffer *buffer;
+} search_state;
+
+static inline void escape_UTF8_char(search_state *state, unsigned char ch_len) {
+    const unsigned char ch = (unsigned char)*state->ptr;
+    switch (ch_len) {
+        case 1: {
+            switch (ch) {
+                case '"':  fbuffer_append(state->buffer, "\\\"", 2); break;
+                case '\\': fbuffer_append(state->buffer, "\\\\", 2); break;
+                case '/':  fbuffer_append(state->buffer, "\\/", 2);  break;
+                case '\b': fbuffer_append(state->buffer, "\\b", 2);  break;
+                case '\f': fbuffer_append(state->buffer, "\\f", 2);  break;
+                case '\n': fbuffer_append(state->buffer, "\\n", 2);  break;
+                case '\r': fbuffer_append(state->buffer, "\\r", 2);  break;
+                case '\t': fbuffer_append(state->buffer, "\\t", 2);  break;
+                default: {
+                    const char *hexdig = "0123456789abcdef";
+                    char scratch[6] = { '\\', 'u', '0', '0', 0, 0 };
+                    scratch[4] = hexdig[(ch >> 4) & 0xf];
+                    scratch[5] = hexdig[ch & 0xf];
+                    fbuffer_append(state->buffer, scratch, 6);
+                    break;
+                }
+            }
+            break;
+        }
+        case 3: {
+            if (state->ptr[2] & 1) {
+                fbuffer_append(state->buffer, "\\u2029", 6);
+            } else {
+                fbuffer_append(state->buffer, "\\u2028", 6);
+            }
+            break;
+        }
+    }
+    state->cursor = (state->ptr += ch_len);
+}
+
+static inline void search_flush(search_state *state)
+{
+    fbuffer_append(state->buffer, state->cursor, state->ptr - state->cursor);
+    state->cursor = state->ptr;
+}
+
+static inline unsigned char search_escape(search_state *state, const unsigned char escape_table[256])
+{
+    while (state->ptr < state->end) {
+        unsigned char ch = (unsigned char)*state->ptr;
+        unsigned char ch_len = escape_table[ch];
+
+        if (RB_UNLIKELY(ch_len)) {
+            if (ch_len & ESCAPE_MASK) {
+                if (RB_UNLIKELY(ch_len == 11)) {
+                    const unsigned char *uptr = (const unsigned char *)state->ptr;
+                    if (!(uptr[1] == 0x80 && (uptr[2] >> 1) == 0x54)) {
+                        state->ptr += 3;
+                        continue;
+                    }
+                }
+                search_flush(state);
```

Comment:
Since `search_escape` is now responsible for flushing bytes that don't need escaping, it should allow to write an SIMD version that does both scanning and copying at once. e.g. load X bytes, and if none need to be escaped, directly copy them.